src/Spectre.libPeakFinder/ExtremaFinder.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * ExtremaFinder.cpp
+ * Finding indices of extrema in signals.
+ *
+ Copyright 2018 Michal Gallus
+ 
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ 
+ http://www.apache.org/licenses/LICENSE-2.0
+ 
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+#include "ExtremaFinder.h"
+#include "Spectre.libStatistics\Math.h"
+#include "Spectre.libFunctional\Filter.h"
+
+namespace spectre::algorithm::peakfinder
+{
+using namespace spectre::statistics::basic_math;
+using namespace spectre::core::functional;
+
+// Branchless cache-efficient zero filter
+static inline Indices GetNonZeros(const Signal& data)
+{
+    const size_t length = data.size();
+    std::vector<unsigned> indices(length + 1);
+    unsigned zeros = 0;
+
+    for (unsigned i = 0; i < length; i++)
+    {
+        bool dataIsZero = data[i] == 0.0;
+        int index = i - zeros + dataIsZero;
+        zeros += dataIsZero;
+
+        indices[index] = i;
+    }
+    indices.resize(length - zeros);
+
+    return indices;
+}
+
+// This function finds indices of intensities that combined together create signal with removed flat lines.
+static inline Indices FilterFlatSegmentIndices(const SignalView& signal)
+{
+    Signal derivatives = differentiate(signal);
+    Indices indicesOfNonZeros = GetNonZeros(derivatives);
+    indicesOfNonZeros.push_back((unsigned)(signal.size() - 1));
```

Comment:
Above you've resized the `indices` to have no place for this element for sure. You'd spare reallocation if you managed to keep its reserved size one element more.

Reply:
So the call
```C++
indices.reserve(length - zeros);
```
Would shrink the vector length but keep the allocated spot for the new push_back?

Reply:
Actually `indices.reserve(length - zeros + 1);` would have enough place allocated for this additional element, but this still results in a reallocation after you shrinked. I'm not sure whether we can achieve it reasonably, I was just throwing an idea ;)